reaper.go

```diff
@@ -119,34 +116,31 @@ func (r *reaper) reap() error {
 // start starts the exec command and registers the process to the reaper.
 // This function is a helper for exec.Cmd.Start() since this needs to be
 // in sync with exec.Cmd.Wait().
-func (r *reaper) start(c *exec.Cmd) error {
+func (r *reaper) start(c *exec.Cmd) (<-chan int, error) {
 	// This lock is very important to avoid any race with reaper.reap().
 	// We don't want the reaper to reap a process before we have added
 	// it to the exit code channel list.
 	r.RLock()
 	defer r.RUnlock()
 
 	if err := c.Start(); err != nil {
-		return err
+		return nil, err
 	}
 
+	exitCodeCh := make(chan int, 1)
+
 	// This channel is buffered so that reaper.reap() will not
 	// block until reaper.wait() listen onto this channel.
-	r.setExitCodeCh(c.Process.Pid, make(chan int, 1))
+	r.setExitCodeCh(c.Process.Pid, exitCodeCh)
 
-	return nil
+	return exitCodeCh, nil
 }
 
 // wait blocks until the expected process has been reaped. After the reaping
 // from the subreaper, the exit code is sent through the provided channel.
 // This function is a helper for exec.Cmd.Wait() and os.Process.Wait() since
 // both cannot be used directly, because of the subreaper.
-func (r *reaper) wait(pid int, proc waitProcess) (int, error) {
-	exitCodeCh, err := r.getExitCodeCh(pid)
-	if err != nil {
-		return -1, err
-	}
-
+func (r *reaper) wait(exitCodeCh chan int, proc waitProcess) (int, error) {
```

Comment:
here is also receiving chan ...

Reply:
and several other places...

Reply:
Wait, I have a problem with that. From `wait()` I get the following error message:
```
./reaper.go:154: invalid operation: close(exitCodeCh) (cannot close receive-only channel)
```
Don't you think we can stay with those channels without detailing the direction ?

Reply:
it is not a good idea to close it in reaper.wait().
if it needs to be closed, the closing operation should be in the reaper loop.

Reply:
If the closing operation happens in the reaper loop, are you sure the exit code will be received even if the `Wait()` is called later ?

Reply:
Nevermind, I have tested it and it's okay, we can actually close the channel after we sent something, the data will be read at some point by the Wait().